src/auth/src/token_cache.rs

```diff
@@ -0,0 +1,351 @@
+        // error/token. Return it.
+        self.current_token().await
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::errors::CredentialError;
+    use crate::token::test::MockTokenProvider;
+    use std::sync::{Arc, Mutex};
+    use std::time::Duration;
+
+    static TOKEN_VALID_DURATION: Duration = Duration::from_secs(3600);
+
+    #[tokio::test]
+    async fn initial_token_success() {
+        let expected = Token {
+            token: "test-token".to_string(),
+            token_type: "Bearer".to_string(),
+            expires_at: None,
+            metadata: None,
+        };
+        let expected_clone = expected.clone();
+
+        let mut mock = MockTokenProvider::new();
+        mock.expect_get_token()
+            .times(1)
+            .return_once(|| Ok(expected_clone));
+
+        let cache = TokenCache::new(mock);
+        let actual = cache.get_token().await.unwrap();
+        assert_eq!(actual, expected);
+
+        // Verify that we use the cached token instead of making a new request
+        // to the mock token provider.
+        let actual = cache.get_token().await.unwrap();
+        assert_eq!(actual, expected);
+    }
+
+    #[tokio::test]
+    async fn initial_token_failure() {
+        let mut mock = MockTokenProvider::new();
+        mock.expect_get_token()
+            .times(2)
+            .returning(|| Err(CredentialError::non_retryable_from_str("fail")));
+
+        let cache = TokenCache::new(mock);
+        assert!(cache.get_token().await.is_err());
+
+        // Verify that a new request is made to the mock token provider when we
+        // don't have a valid token.
+        assert!(cache.get_token().await.is_err());
+    }
+
+    #[tokio::test(start_paused = true)]
+    async fn expired_token_success() {
+        let now = Instant::now();
+
+        let initial = Token {
+            token: "initial-token".to_string(),
+            token_type: "Bearer".to_string(),
+            expires_at: Some((now + TOKEN_VALID_DURATION).into_std()),
+            metadata: None,
+        };
+        let initial_clone = initial.clone();
+
+        let refresh = Token {
+            token: "refresh-token".to_string(),
+            token_type: "Bearer".to_string(),
+            expires_at: Some((now + 2 * TOKEN_VALID_DURATION).into_std()),
+            metadata: None,
+        };
+        let refresh_clone = refresh.clone();
+
+        let mut mock = MockTokenProvider::new();
+        mock.expect_get_token()
+            .times(1)
+            .return_once(|| Ok(initial_clone));
+
+        mock.expect_get_token()
+            .times(1)
+            .return_once(|| Ok(refresh_clone));
+
+        // fetch an initial token
+        let cache = TokenCache::new(mock);
+        let actual = cache.get_token().await.unwrap();
+        assert_eq!(actual, initial);
+
+        // wait long enough for the token to be expired
+        let sleep = TOKEN_VALID_DURATION;
```

Comment:
we are sleeping for an hour in the test?

I guess I am missing something here.

Reply:
`tokio::time::Instant` is like a wrapper over `std::time::Instant` that lets us fake the time in a test.

So the code is more like "pretend to sleep for an hour".

https://docs.rs/tokio/latest/tokio/time/fn.advance.html

Reply:
thats awesome!